File: coreza-backend/function/dhan.py
import os
import uuid
from typing import Any, Dict, List, Tuple
from datetime import datetime, timedelta, timezone
import csv
from functools import lru_cache

from dotenv import load_dotenv
from config.supabaseClient import create_client
import httpx
from fastapi import HTTPException, status
from pydantic import BaseModel

# Load environment variables
load_dotenv()
SUPABASE_URL = os.getenv("SUPABASE_URL")
SUPABASE_KEY = os.getenv("SUPABASE_KEY")
if not SUPABASE_URL or not SUPABASE_KEY:
    raise RuntimeError("SUPABASE_URL and SUPABASE_KEY must be set as environment variables.")

# Initialize Supabase client
supabase = create_client(SUPABASE_URL, SUPABASE_KEY)

# Base URL and Scrip-Master URL for Dhan API
BASE_URL = "https://sandbox.dhan.co/v2"
SCRIP_MASTER_URL = "https://images.dhan.co/api-data/api-scrip-master-detailed.csv"

# --- Request models ---
class AuthRequest(BaseModel):
    user_id: str
    credential_name: str
    client_id: str
    api_key: str

class CancelOrderRequest(BaseModel):
    user_id: str
    credential_id: str
    order_id: str

class GetCandleRequest(BaseModel):
    user_id: str
    credential_id: str
    exchange: str
    symbol: str
    interval: str  # '1Min', '5Min', '15Min', '1Hour', '1Day'
    lookback: int

# --- Helper to fetch credentials ---
def get_api_credentials(user_id: str, credential_id: str) -> Dict[str, str]:
    try:
        res = supabase.table("user_credentials") \
            .select("client_json") \
            .eq("user_id", user_id) \
            .eq("name", credential_id) \
            .eq("service_type", "dhan") \
            .single() \
            .execute()
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Supabase error: {e}")
    creds = res.data.get("client_json", {})
    client_id = creds.get("client_id")
    api_key = creds.get("api_key")
    if not client_id or not api_key:
        raise HTTPException(status_code=400, detail="API credentials not found.")
    return {"client_id": client_id, "api_key": api_key}

# --- Load and lookup Scrip-Master CSV ---
@lru_cache(maxsize=1)
def load_scrip_master() -> Dict[Tuple[str, str], str]:
    resp = httpx.get(SCRIP_MASTER_URL, timeout=30)
    try:
        resp.raise_for_status()
    except Exception as e:
        raise HTTPException(status_code=status.HTTP_502_BAD_GATEWAY, detail=f"Failed to load scrip master: {e}")
    reader = csv.DictReader(resp.text.splitlines())
    master: Dict[Tuple[str, str], str] = {}
    for row in reader:
        seg = row.get("EXCH_ID")  # e.g., 'NSE_EQ'
        sym = row.get("UNDERLYING_SYMBOL")  # human-friendly symbol
        secid = row.get("SECURITY_ID")  # numeric ID
        if seg and sym and secid:
            master[(seg, sym.upper())] = secid
    return master


def lookup_security_id(symbol: str, exchange_segment: str) -> str:
    master = load_scrip_master()
    key = (exchange_segment, symbol.upper())
    try:
        return master[key]
    except KeyError:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=f"Security ID not found for symbol {symbol} on segment {exchange_segment}"
        )

# --- Service-layer functions ---
def auth_url_func(request: AuthRequest) -> Dict[str, Any]:
    headers = {"access-token": request.api_key, "Accept": "application/json"}
    try:
        resp = httpx.get(f"{BASE_URL}/user/profile", headers=headers, timeout=10)
    except httpx.RequestError as exc:
        raise HTTPException(status_code=status.HTTP_502_BAD_GATEWAY, detail=f"Network error: {exc}")
    if resp.status_code != 200:
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail="Validation failed: " + resp.text)
    cred_id = str(uuid.uuid4())
    record = {
        "id": cred_id,
        "user_id": request.user_id,
        "name": request.credential_name,
        "service_type": "dhan",
        "client_json": {"client_id": request.client_id, "api_key": request.api_key}
    }
    upsert = supabase.table("user_credentials").upsert(record, on_conflict="user_id,name").execute()
    if not upsert.data:
        raise HTTPException(status_code=500, detail="Failed to store credentials.")
    return {"success": True, "credential_id": upsert.data[0].get("id", cred_id)}

def list_credentials_func(user_id: str) -> List[Dict[str, Any]]:
    resp = supabase.table("user_credentials")\
        .select("id,name")\
        .eq("user_id", user_id)\
        .eq("service_type", "dhan")\
        .execute()
    return resp.data

def get_account_func(params: Dict[str, Any]) -> Dict[str, Any]:
    user_id = params.get("user_id")
    credential_id = params.get("credential_id")
    creds = get_api_credentials(user_id, credential_id)
    headers = {"access-token": creds["api_key"], "Accept": "application/json"}
    resp = httpx.get(f"{BASE_URL}/orders/account", headers=headers, timeout=10)
    if resp.status_code != 200:
        raise HTTPException(status_code=status.HTTP_502_BAD_GATEWAY, detail="Dhan API error: " + resp.text)
    return resp.json()

def list_positions_func(params: Dict[str, Any]) -> Dict[str, Any]:
    user_id = params.get("user_id")
    credential_id = params.get("credential_id")
    creds = get_api_credentials(user_id, credential_id)
    headers = {"access-token": creds["api_key"], "Accept": "application/json"}
    resp = httpx.get(f"{BASE_URL}/orders/holdings", headers=headers, timeout=10)
    if resp.status_code != 200:
        raise HTTPException(status_code=status.HTTP_502_BAD_GATEWAY, detail="Dhan API error: " + resp.text)
    return resp.json()

def list_orders_func(params: Dict[str, Any]) -> Dict[str, Any]:
    user_id = params.get("user_id")
    credential_id = params.get("credential_id")
    creds = get_api_credentials(user_id, credential_id)
    headers = {"access-token": creds["api_key"], "Accept": "application/json"}
    resp = httpx.get(f"{BASE_URL}/order", headers=headers, timeout=10)
    if resp.status_code != 200:
        raise HTTPException(status_code=status.HTTP_502_BAD_GATEWAY, detail="Dhan API error: " + resp.text)
    return resp.json()

def cancel_order_func(request: CancelOrderRequest) -> Dict[str, Any]:
    creds = get_api_credentials(request.user_id, request.credential_id)
    headers = {"access-token": creds["api_key"], "Accept": "application/json"}
    payload = {"orderId": request.order_id}
    resp = httpx.post(f"{BASE_URL}/order/cancel", headers=headers, json=payload, timeout=10)
    if resp.status_code != 200:
        raise HTTPException(status_code=status.HTTP_502_BAD_GATEWAY, detail="Dhan API error: " + resp.text)
    return resp.json()

def get_candle_func(request: GetCandleRequest) -> Dict[str, Any]:
    """
    Fetch historical or intraday bar data from Dhan API using the proper securityId.
    Formats the returned bars into a standardized candle list.
    """
    # Extract and validate parameters
    user_id = request["user_id"]
    exchange = request["exchange"]
    credential_id = request["credential_id"]
    symbol = request["symbol"]
    interval = request["interval"]
    lookback = request["lookback"]

    if not all([user_id, credential_id, exchange, symbol, interval, lookback]):
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail="Missing parameters for get_candle")

    # Retrieve API credentials and prepare headers
    creds = get_api_credentials(user_id, credential_id)
    api_key = creds["api_key"]
    headers = {
        "access-token": api_key,
        "Accept": "application/json",
        "Content-Type": "application/json"
    }
    today = datetime.now(timezone.utc).date()

    # Map symbol to security ID
    security_id = lookup_security_id(symbol, exchange.split("_")[0])

    # Build payload
    endpoint = f"{BASE_URL}/charts/historical" if interval == "1Day" else f"{BASE_URL}/charts/intraday"
    payload: Dict[str, Any] = {
        "securityId": security_id,
        "exchangeSegment": exchange,
        "instrument": "EQUITY",
        "expiryCode": 0,
        "oi": False
    }
    if interval == "1Day":
        payload.update({
            "fromDate": (today - timedelta(days=lookback)).strftime("%Y-%m-%d"),
            "toDate": today.strftime("%Y-%m-%d")
        })

    # Fetch data
    try:
        resp = httpx.post(endpoint, headers=headers, json=payload, timeout=10)
    except httpx.RequestError as exc:
        raise HTTPException(status_code=status.HTTP_502_BAD_GATEWAY, detail=f"Network error: {exc}")

    if resp.status_code != 200:
        try:
            err = resp.json()
            detail_msg = err.get("errorMessage") or resp.text
        except Exception:
            detail_msg = resp.text
        raise HTTPException(status_code=status.HTTP_502_BAD_GATEWAY, detail="Dhan API error: " + detail_msg)

    result = resp.json()
    # Handle bars returned as dict of lists (open, high, low, close, volume, timestamp)
    bars = result.get("data", result)
    if isinstance(bars, dict) and isinstance(bars.get("open"), list):
        length = len(bars.get("open"))
        reconstructed: List[Dict[str, Any]] = []
        for i in range(length):
            reconstructed.append({
                "timestamp": bars.get("timestamp")[i],
                "open": bars.get("open")[i],
                "high": bars.get("high")[i],
                "low": bars.get("low")[i],
                "close": bars.get("close")[i],
                "volume": bars.get("volume")[i]
            })
        bars = reconstructed
    # Trim to lookback if needed
    if interval != "1Day" and isinstance(bars, list):
        bars = bars[-lookback:]
    if interval != "1Day" and isinstance(bars, list):
        bars = bars[-lookback:]

    # Transform bars into candle dicts matching Alpaca format
    candles: List[Dict[str, Any]] = []
    for bar in bars:
        # Skip invalid entries
        if not isinstance(bar, dict):
            continue
        # Extract timestamp
        ts = bar.get("timestamp") or bar.get("t") or bar.get("T")
        if not isinstance(ts, str) and hasattr(ts, "isoformat"):
            try:
                ts = ts.isoformat()
            except Exception:
                ts = str(ts)
        # Extract OHLCV values
        o = bar.get("open") or bar.get("o") or bar.get("O")
        h = bar.get("high") or bar.get("h") or bar.get("H")
        l = bar.get("low") or bar.get("l") or bar.get("L")
        c = bar.get("close") or bar.get("c") or bar.get("C")
        v = bar.get("volume") or bar.get("v") or bar.get("V")
        candles.append({"t": ts, "o": o, "h": h, "l": l, "c": c, "v": v})

    return {"symbol": symbol, "candles": candles}

File: coreza-backend/src/routes/dhan.ts
import express from 'express';
import { BrokerService } from '../services/brokers';

const router = express.Router();

// Get user credentials list for Dhan
router.get('/credentials', async (req, res, next) => {
  try {
    const { user_id } = req.query;
    
    if (!user_id) {
      return res.status(400).json({ error: 'user_id is required' });
    }
    
    const credentials = await BrokerService.getCredentialsList('dhan', user_id as string);
    
    res.json({
      success: true,
      credentials
    });
  } catch (error) {
    next(error);
  }
});

// Add auth-url endpoint for authAction
router.post('/auth-url', async (req, res, next) => {
  try {
    const { user_id, credential_name, client_id, api_key } = req.body;
    
    if (!user_id || !credential_name || !client_id || !api_key) {
      return res.status(400).json({ error: 'Missing required parameters' });
    }

    const data = await BrokerService.saveCredentials('dhan', user_id, credential_name, {
      api_key,
      client_id
    });

    res.json({
      success: true,
      message: 'Dhan credentials saved successfully',
      credential_id: data.id
    });
  } catch (error) {
    next(error);
  }
});

// Dynamic operation endpoint to match node pattern
router.post('/:operation', async (req, res, next) => {
  try {
    const { operation } = req.params;
    const { user_id, credential_id } = req.body;
    
    if (!user_id || !credential_id) {
      return res.status(400).json({ error: 'user_id and credential_id are required' });
    }

    const result = await BrokerService.execute('dhan', {
      user_id,
      credential_id,
      operation,
      ...req.body
    });

    if (!result.success) {
      return res.status(400).json({ error: result.error });
    }

    res.json(result.data);
  } catch (error) {
    next(error);
  }
});

export default router;

// Dhan API Configuration
const BASE_URL = 'https://sandbox.dhan.co/v2';
const SCRIP_MASTER_URL = 'https://images.dhan.co/api-data/api-scrip-master-detailed.csv';

// Cache for scrip master data
let scripMasterCache: Map<string, string> | null = null;
let scripMasterCacheTime = 0;
const CACHE_DURATION = 3600000; // 1 hour in milliseconds

interface DhanCredentials {
  client_id: string;
  api_key: string;
}

// Helper function to get API credentials
async function getApiCredentials(userId: string, credentialId: string): Promise<DhanCredentials> {
  try {
    const { data, error } = await supabase
      .from('user_credentials')
      .select('client_json')
      .eq('user_id', userId)
      .eq('name', credentialId)
      .eq('service_type', 'dhan')
      .single();

    if (error || !data) {
      throw createError('Dhan credentials not found', 404);
    }

    const creds = data.client_json;
    if (!creds.client_id || !creds.api_key) {
      throw createError('Invalid Dhan API credentials', 400);
    }

    return {
      client_id: creds.client_id,
      api_key: creds.api_key
    };
  } catch (error) {
    throw createError('Failed to retrieve Dhan credentials', 500);
  }
}

// Load scrip master data with caching
async function loadScripMaster(): Promise<Map<string, string>> {
  const now = Date.now();
  
  if (scripMasterCache && (now - scripMasterCacheTime) < CACHE_DURATION) {
    return scripMasterCache;
  }

  try {
    const response = await axios.get(SCRIP_MASTER_URL, { timeout: 30000 });
    const csvData = response.data;
    
    const lines = csvData.trim().split('\n');
    const headers = lines[0].split(',');
    
    const exch_id_index = headers.indexOf('EXCH_ID');
    const symbol_index = headers.indexOf('UNDERLYING_SYMBOL');
    const security_id_index = headers.indexOf('SECURITY_ID');
    
    if (exch_id_index === -1 || symbol_index === -1 || security_id_index === -1) {
      throw new Error('Required columns not found in scrip master CSV');
    }

    const master = new Map<string, string>();
    
    for (let i = 1; i < lines.length; i++) {
      const columns = lines[i].split(',');
      const exch_id = columns[exch_id_index]?.trim();
      const symbol = columns[symbol_index]?.trim();
      const security_id = columns[security_id_index]?.trim();
      
      if (exch_id && symbol && security_id) {
        const key = `${exch_id}:${symbol.toUpperCase()}`;
        master.set(key, security_id);
      }
    }

    scripMasterCache = master;
    scripMasterCacheTime = now;
    
    console.log(`Loaded ${master.size} securities from Dhan scrip master`);
    return master;
  } catch (error) {
    console.error('Failed to load scrip master:', error);
    throw createError('Failed to load Dhan scrip master data', 502);
  }
}

// Lookup security ID for symbol
async function lookupSecurityId(symbol: string, exchangeSegment: string): Promise<string> {
  const master = await loadScripMaster();
  const key = `${exchangeSegment}:${symbol.toUpperCase()}`;
  const securityId = master.get(key);
  
  if (!securityId) {
    throw createError(`Security ID not found for ${symbol} on ${exchangeSegment}`, 404);
  }
  
  return securityId;
}

// Save credentials
router.post('/auth', async (req, res, next) => {
  try {
    const { user_id, credential_name, client_id, api_key } = req.body;
    
    if (!user_id || !credential_name || !client_id || !api_key) {
      throw createError('Missing required parameters', 400);
    }

    // Test the credentials by making a test API call
    const headers = {
      'access-token': api_key,
      'Accept': 'application/json'
    };

    try {
      const testResponse = await axios.get(`${BASE_URL}/funds`, { headers, timeout: 10000 });
      
      if (testResponse.status !== 200) {
        throw createError('Invalid Dhan API credentials', 401);
      }
    } catch (error) {
      if (axios.isAxiosError(error) && error.response?.status === 401) {
        throw createError('Invalid Dhan API credentials', 401);
      }
      throw createError('Failed to validate Dhan credentials', 502);
    }

    // Save credentials to database
    const { data, error } = await supabase
      .from('user_credentials')
      .upsert({
        user_id,
        name: credential_name,
        service_type: 'dhan',
        client_json: { client_id, api_key }
      })
      .select()
      .single();

    if (error) {
      throw createError('Failed to save credentials', 500);
    }

    res.json({
      message: 'Dhan credentials saved successfully',
      credential_id: data.id
    });
  } catch (error) {
    next(error);
  }
});

// Get account funds
router.post('/funds', async (req, res, next) => {
  try {
    const { user_id, credential_id } = req.body;
    
    if (!user_id || !credential_id) {
      throw createError('user_id and credential_id are required', 400);
    }

    const creds = await getApiCredentials(user_id, credential_id);
    
    const headers = {
      'access-token': creds.api_key,
      'Accept': 'application/json'
    };

    const response = await axios.get(`${BASE_URL}/funds`, { headers, timeout: 10000 });
    
    res.json(response.data);
  } catch (error) {
    if (axios.isAxiosError(error) && error.response) {
      throw createError(`Dhan API error: ${error.response.data}`, error.response.status);
    }
    next(error);
  }
});

// Get positions
router.post('/positions', async (req, res, next) => {
  try {
    const { user_id, credential_id } = req.body;
    
    if (!user_id || !credential_id) {
      throw createError('user_id and credential_id are required', 400);
    }

    const creds = await getApiCredentials(user_id, credential_id);
    
    const headers = {
      'access-token': creds.api_key,
      'Accept': 'application/json'
    };

    const response = await axios.get(`${BASE_URL}/positions`, { headers, timeout: 10000 });
    
    res.json(response.data);
  } catch (error) {
    if (axios.isAxiosError(error) && error.response) {
      throw createError(`Dhan API error: ${error.response.data}`, error.response.status);
    }
    next(error);
  }
});

// Get holdings
router.post('/holdings', async (req, res, next) => {
  try {
    const { user_id, credential_id } = req.body;
    
    if (!user_id || !credential_id) {
      throw createError('user_id and credential_id are required', 400);
    }

    const creds = await getApiCredentials(user_id, credential_id);
    
    const headers = {
      'access-token': creds.api_key,
      'Accept': 'application/json'
    };

    const response = await axios.get(`${BASE_URL}/holdings`, { headers, timeout: 10000 });
    
    res.json(response.data);
  } catch (error) {
    if (axios.isAxiosError(error) && error.response) {
      throw createError(`Dhan API error: ${error.response.data}`, error.response.status);
    }
    next(error);
  }
});

// Get orders
router.post('/orders', async (req, res, next) => {
  try {
    const { user_id, credential_id } = req.body;
    
    if (!user_id || !credential_id) {
      throw createError('user_id and credential_id are required', 400);
    }

    const creds = await getApiCredentials(user_id, credential_id);
    
    const headers = {
      'access-token': creds.api_key,
      'Accept': 'application/json'
    };

    const response = await axios.get(`${BASE_URL}/order`, { headers, timeout: 10000 });
    
    res.json(response.data);
  } catch (error) {
    if (axios.isAxiosError(error) && error.response) {
      throw createError(`Dhan API error: ${error.response.data}`, error.response.status);
    }
    next(error);
  }
});

// Cancel order
router.post('/orders/cancel', async (req, res, next) => {
  try {
    const { user_id, credential_id, order_id } = req.body;
    
    if (!user_id || !credential_id || !order_id) {
      throw createError('user_id, credential_id, and order_id are required', 400);
    }

    const creds = await getApiCredentials(user_id, credential_id);
    
    const headers = {
      'access-token': creds.api_key,
      'Accept': 'application/json',
      'Content-Type': 'application/json'
    };

    const payload = { orderId: order_id };

    const response = await axios.post(`${BASE_URL}/order/cancel`, payload, { headers, timeout: 10000 });
    
    res.json(response.data);
  } catch (error) {
    if (axios.isAxiosError(error) && error.response) {
      throw createError(`Dhan API error: ${error.response.data}`, error.response.status);
    }
    next(error);
  }
});

// Get historical/intraday candle data
router.post('/candles', async (req, res, next) => {
  try {
    const { user_id, credential_id, exchange, symbol, interval, lookback } = req.body;
    
    if (!user_id || !credential_id || !exchange || !symbol || !interval || !lookback) {
      throw createError('All parameters are required: user_id, credential_id, exchange, symbol, interval, lookback', 400);
    }

    const creds = await getApiCredentials(user_id, credential_id);
    
    // Get security ID from scrip master
    const exchangeSegment = exchange.split('_')[0]; // Extract NSE from NSE_EQ
    const securityId = await lookupSecurityId(symbol, exchange);
    
    const headers = {
      'access-token': creds.api_key,
      'Accept': 'application/json',
      'Content-Type': 'application/json'
    };

    // Determine endpoint and payload based on interval
    const isDaily = interval === '1Day';
    const endpoint = isDaily 
      ? `${BASE_URL}/charts/historical` 
      : `${BASE_URL}/charts/intraday`;

    const payload: any = {
      securityId,
      exchangeSegment: exchange,
      instrument: 'EQUITY',
      expiryCode: 0,
      oi: false
    };

    if (isDaily) {
      const today = new Date();
      const fromDate = new Date(today);
      fromDate.setDate(today.getDate() - lookback);
      
      payload.fromDate = fromDate.toISOString().split('T')[0];
      payload.toDate = today.toISOString().split('T')[0];
    }

    const response = await axios.post(endpoint, payload, { headers, timeout: 10000 });
    
    let bars = response.data.data || response.data;
    
    // Handle different response formats
    if (typeof bars === 'object' && Array.isArray(bars.open)) {
      // Format: { open: [...], high: [...], low: [...], close: [...], volume: [...], timestamp: [...] }
      const length = bars.open.length;
      const reconstructed = [];
      
      for (let i = 0; i < length; i++) {
        reconstructed.push({
          timestamp: bars.timestamp[i],
          open: bars.open[i],
          high: bars.high[i],
          low: bars.low[i],
          close: bars.close[i],
          volume: bars.volume[i]
        });
      }
      bars = reconstructed;
    }

    // Limit to lookback for intraday data
    if (!isDaily && Array.isArray(bars)) {
      bars = bars.slice(-lookback);
    }

    // Transform to standardized candle format
    const candles = bars.map((bar: any) => {
      const timestamp = bar.timestamp || bar.t || bar.T;
      const open = bar.open || bar.o || bar.O;
      const high = bar.high || bar.h || bar.H;
      const low = bar.low || bar.l || bar.L;
      const close = bar.close || bar.c || bar.C;
      const volume = bar.volume || bar.v || bar.V;
      
      return {
        t: timestamp,
        o: open,
        h: high,
        l: low,
        c: close,
        v: volume
      };
    });

    res.json({
      symbol,
      exchange,
      interval,
      candles
    });
  } catch (error) {
    if (axios.isAxiosError(error) && error.response) {
      throw createError(`Dhan API error: ${error.response.data}`, error.response.status);
    }
    next(error);
  }
});

// Get quote data
router.post('/quote', async (req, res, next) => {
  try {
    const { user_id, credential_id, exchange, symbol } = req.body;
    
    if (!user_id || !credential_id || !exchange || !symbol) {
      throw createError('user_id, credential_id, exchange, and symbol are required', 400);
    }

    const creds = await getApiCredentials(user_id, credential_id);
    const securityId = await lookupSecurityId(symbol, exchange);
    
    const headers = {
      'access-token': creds.api_key,
      'Accept': 'application/json',
      'Content-Type': 'application/json'
    };

    const payload = {
      securityId,
      exchangeSegment: exchange,
      instrument: 'EQUITY'
    };

    const response = await axios.post(`${BASE_URL}/marketfeed/ltp`, payload, { headers, timeout: 10000 });
    
    res.json({
      symbol,
      exchange,
      data: response.data
    });
  } catch (error) {
    if (axios.isAxiosError(error) && error.response) {
      throw createError(`Dhan API error: ${error.response.data}`, error.response.status);
    }
    next(error);
  }
});

export default router;